compile failure: isinstance not supported

``` python
import numpy as np

def conj (x):
    if x.dtype == complex:
        return x.real - 1j*x.imag
    else:
        return x

#pythran export mag_sqr(float64[])
#pythran export mag_sqr(complex128[])
def mag_sqr (x):
    return x * conj(x)
```

---

/tmp/tmpVJcDTS.cpp:39:183: error: no matching function for call to  getattr(pythonic::types::ndarray<double, 1ul>) 
       typedef decltype(pythonic::**builtin**::getattrpythonic::types::attr::REAL(std::declval<typename std::remove_cv<typename std::remove_reference<argument_type0>::type>::type>())) __type0;

--- a/pythran/analyses/cfg.py
+++ b/pythran/analyses/cfg.py
@@ -36,6 +36,7 @@ class CFG(FunctionAnalysis):
         super(CFG, self).__init__()
 
     def visit_FunctionDef(self, node):
+        """OUT = node, RAISES = ()"""
         # the function itself is the entry point
         self.result.add_node(node)
         currs = (node,)
@@ -49,6 +50,7 @@ class CFG(FunctionAnalysis):
         self.result.add_node(CFG.NIL)
         for curr in currs:
             self.result.add_edge(curr, CFG.NIL)
+        return (node,), ()
 
     def visit_Pass(self, node):
         """OUT = node, RAISES = ()"""

--- a/pythran/analyses/static_expressions.py
+++ b/pythran/analyses/static_expressions.py
@@ -60,7 +60,7 @@ class StaticExpressions(NodeAnalysis):
         return self.visit(node.func)and self.add(node)  # very limited
 
     def visit_Attribute(self, node):
-        return node.attr == 'is_none'
+        return node.attr in ('is_none', 'isinstance')
 
     def visit_Constant(self, node):
         self.constant_expressions.add(node)

--- a/pythran/pythonic/__builtin__/isinstance.hpp
+++ b/pythran/pythonic/__builtin__/isinstance.hpp
@@ -1,0 +1,7 @@
+#ifndef PYTHONIC_BUILTIN_ISINSTANCE_HPP
+#define PYTHONIC_BUILTIN_ISINSTANCE_HPP
+
+#include "pythonic/include/__builtin__/isinstance.hpp"
+#include "pythonic/include/__builtin__/pythran/is_none.hpp"
+
+#endif

--- a/pythran/pythonic/include/__builtin__/isinstance.hpp
+++ b/pythran/pythonic/include/__builtin__/isinstance.hpp
@@ -1,0 +1,43 @@
+#ifndef PYTHONIC_INCLUDE_BUILTIN_ISINSTANCE_HPP
+#define PYTHONIC_INCLUDE_BUILTIN_ISINSTANCE_HPP
+
+#include "pythonic/include/utils/functor.hpp"
+#include "pythonic/__builtin__/pythran/is_none.hpp"
+
+PYTHONIC_NS_BEGIN
+namespace types
+{
+  struct str;
+
+  template <class Ty0, class Ty1>
+  struct isinstance
+      : std::conditional<std::is_same<Ty0, Ty1>::value, true_type, false_type> {
+  };
+
+  // some specialization
+
+  template <>
+  struct isinstance<char const *, str> {
+    using type = true_type;
+  };
+  template <>
+  struct isinstance<str, char const *> {
+    using type = true_type;
+  };
+}
+
+namespace __builtin__
+{
+  template <class Obj, class Cls>
+  typename types::isinstance<
+      Obj, typename std::decay<decltype(std::declval<Cls>()())>::type>::type
+      isinstance(Obj, Cls)
+  {
+    return {};
+  }
+
+  DEFINE_FUNCTOR(pythonic::__builtin__, isinstance);
+}
+PYTHONIC_NS_END
+
+#endif

--- a/pythran/pythonic/include/__builtin__/pythran/is_none.hpp
+++ b/pythran/pythonic/include/__builtin__/pythran/is_none.hpp
@@ -115,6 +115,12 @@ namespace __builtin__
     DEFINE_FUNCTOR(pythonic::__builtin__::pythran, is_none);
   }
 }
+template <>
+struct to_python<types::true_type> : to_python<bool> {
+};
+template <>
+struct to_python<types::false_type> : to_python<bool> {
+};
 PYTHONIC_NS_END
 
 #endif

--- a/pythran/pythonic/include/types/bool.hpp
+++ b/pythran/pythonic/include/types/bool.hpp
@@ -10,6 +10,7 @@ template <>
 struct to_python<bool> {
   static PyObject *convert(bool b);
 };
+
 template <>
 struct from_python<bool> {
   static bool is_convertible(PyObject *obj);

--- a/pythran/tables.py
+++ b/pythran/tables.py
@@ -2702,6 +2702,7 @@ MODULES = {
         "hex": ConstFunctionIntr(signature=Fun[[int], str]),
         "id": ConstFunctionIntr(signature=Fun[[T0], int]),
         "int": ConstFunctionIntr(signature=_int_signature),
+        "isinstance": ConstFunctionIntr(signature=Fun[[T0, T1], bool]),
         "iter": FunctionIntr(
             signature=Fun[[Iterable[T0]], Generator[T0]]),  # not const
         "len": ConstFunctionIntr(

--- a/pythran/tests/test_none.py
+++ b/pythran/tests/test_none.py
@@ -1,7 +1,8 @@
 from pythran.tests import TestEnv
 from unittest import skip
-from pythran.typing import List, Dict
+from pythran.typing import List, Dict, NDArray
 import pythran
+import numpy as np
 
 class TestNone(TestEnv):
 
@@ -567,3 +568,56 @@ def returned_none_member(a):
                     diameter = 20.0
                 return diameter'''
         self.run_test(code, {'a': '10'}, none_escaping1=[Dict[str,str]])
+
+
+class TestIsInstance(TestEnv):
+
+    def test_isinstance_int0(self):
+        self.run_test(
+            'def isinstance_int0(x, y): return isinstance(x, int), isinstance(y, int)',
+            1, '1',
+            isinstance_int0=[int, str])
+
+    def test_isinstance_int1(self):
+        code = 'def isinstance_int1a(x):\n if isinstance(x, int): return x  + 1\n else: return  x * 3'
+        self.run_test(code, 1, isinstance_int1a=[int])
+        code = 'def isinstance_int1b(x):\n if isinstance(x, int): return x  + 1\n else: return  x * 3'
+        self.run_test(code, '1', isinstance_int1b=[str])
+
+    def test_isinstance_int2(self):
+        code = 'def isinstance_int2(x):\n if isinstance(x, int): return 1\n else: return  "3"'
+        self.run_test(code, 1, isinstance_int2=[int])
+
+    def test_isinstance_int3(self):
+        code = 'def isinstance_int3(x):\n if isinstance(x, int) or x is None: return 1\n else: return  "3"'
+        self.run_test(code, 1, isinstance_int3=[int])
+
+    def test_isinstance_complex0(self):
+        code = '''
+            import numpy as np
+
+            def conj (x):
+                if isinstance(x.dtype, complex):
+                    return x.real - 1j*x.imag
+                else:
+                    return x
+
+            def isinstance_complex0(x):
+                return x * conj(x)'''
+        self.run_test(code, np.ones(5, dtype=int),
+                      isinstance_complex0=[NDArray[int,:]])
+
+    def test_isinstance_complex1(self):
+        code = '''
+            import numpy as np
+
+            def conj (x):
+                if isinstance(x.dtype, complex):
+                    return x.real - 1j*x.imag
+                else:
+                    return x
+
+            def isinstance_complex1(x):
+                return x * conj(x)'''
+        self.run_test(code, np.ones(5, dtype=complex) * 2j,
+                      isinstance_complex1=[NDArray[complex,:]])

--- a/pythran/transformations/normalize_static_if.py
+++ b/pythran/transformations/normalize_static_if.py
@@ -1,6 +1,6 @@
 """ NormalizeStaticIf adds support for static guards. """
 
-from pythran.analyses import (ImportedIds, HasReturn, IsAssigned,
+from pythran.analyses import (ImportedIds, HasReturn, IsAssigned, CFG,
                               HasBreak, HasContinue, DefUseChains, Ancestors,
                               StaticExpressions, HasStaticExpression)
 from pythran.passmanager import Transformation
@@ -125,6 +125,7 @@ class NormalizeStaticIf(Transformation):
     def visit_Module(self, node):
         self.new_functions = []
         self.funcs = []
+        self.cfgs = []
         self.generic_visit(node)
         node.body.extend(self.new_functions)
         return node
@@ -179,9 +180,11 @@ class NormalizeStaticIf(Transformation):
         return "$isstatic{}".format(len(self.new_functions) + 1)
 
     def visit_FunctionDef(self, node):
+        self.cfgs.append(self.gather(CFG, node))
         self.funcs.append(node)
         onode = self.generic_visit(node)
         self.funcs.pop()
+        self.cfgs.pop()
         return onode
 
     def visit_IfExp(self, node):
@@ -287,23 +290,33 @@ class NormalizeStaticIf(Transformation):
         cont_n = "$cont{}".format(n)
 
         if has_return:
+            cfg = self.cfgs[-1]
+            always_return = all(isinstance(x, (ast.Return, ast.Yield))
+                                for x in cfg[node])
+            always_return &= true_has_return and false_has_return
 
-            cont_ass = self.make_control_flow_handlers(cont_n, status_n,
-                                                       expected_return,
-                                                       has_cont, has_break)
-
-            cmpr = ast.Compare(ast.Name(status_n, ast.Load(), None, None),
-                               [ast.Eq()], [ast.Constant(EARLY_RET, None)])
 
             fast_return = [ast.Name(status_n, ast.Store(), None, None),
                            ast.Name(return_n, ast.Store(), None, None),
                            ast.Name(cont_n, ast.Store(), None, None)]
 
-            return [ast.Assign([ast.Tuple(fast_return, ast.Store())],
-                               actual_call),
-                    ast.If(cmpr,
-                           [ast.Return(ast.Name(return_n, ast.Load(), None, None))],
-                           cont_ass)]
+            if always_return:
+                return [ast.Assign([ast.Tuple(fast_return, ast.Store())],
+                                   actual_call),
+                        ast.Return(ast.Name(return_n, ast.Load(), None, None))]
+            else:
+                cont_ass = self.make_control_flow_handlers(cont_n, status_n,
+                                                           expected_return,
+                                                           has_cont, has_break)
+
+                cmpr = ast.Compare(ast.Name(status_n, ast.Load(), None, None),
+                                   [ast.Eq()], [ast.Constant(EARLY_RET, None)])
+                return [ast.Assign([ast.Tuple(fast_return, ast.Store())],
+                                   actual_call),
+                        ast.If(cmpr,
+                               [ast.Return(ast.Name(return_n, ast.Load(),
+                                                    None, None))],
+                               cont_ass)]
         elif has_break or has_cont:
             cont_ass = self.make_control_flow_handlers(cont_n, status_n,
                                                        expected_return,

